[Bug]: Crash on batch market instruction

### Problem encountered

Market sim fuzzing run is hitting a crashing bug when sending a batch market transaction


### Observed behaviour

```
panic: runtime error: index out of range [9] with length 9

goroutine 38333 [running]:
code.vegaprotocol.io/vega/core/processor.(*BMIProcessor).ProcessBatch(0xc00e8bd8f8, {0x428ba40, 0xc009125e90}, 0xc00226a960, {0xc009264a80, 0x40}, {0xc00af7d980, 0x40}, {0x42d0908, 0xc000d62240})
	/jenkins/workspace/common/vega-market-sim-reinforcement/extern/vega/core/processor/batch_market_instructions_processor.go:245 +0x1675
code.vegaprotocol.io/vega/core/processor.(*App).DeliverBatchMarketInstructions(0xc000238e00, {0x428ba40, 0xc009125e90}, {0x42b5440, 0xc00bf51280}, {0xc00af7d980, 0x40})
	/jenkins/workspace/common/vega-market-sim-reinforcement/extern/vega/core/processor/abci.go:1437 +0x16a
code.vegaprotocol.io/vega/core/processor.addDeterministicID.func1({0x428ba40, 0xc009125e90}, {0x42b5440, 0xc00bf51280})
	/jenkins/workspace/common/vega-market-sim-reinforcement/extern/vega/core/processor/abci.go:508 +0x103
code.vegaprotocol.io/vega/core/processor.(*App).CheckBatchMarketInstructionsW.func1({0x428ba40, 0xc009125e90}, {0x42b5440, 0xc00bf51280})
	/jenkins/workspace/common/vega-market-sim-reinforcement/extern/vega/core/processor/abci.go:530 +0x76
code.vegaprotocol.io/vega/core/processor.(*App).SendTransactionResult.func1({0x428ba40, 0xc009125e90}, {0x42b5440, 0xc00bf51280})
	/jenkins/workspace/common/vega-market-sim-reinforcement/extern/vega/core/processor/abci.go:560 +0x5c
```

### Expected behaviour

Not crashing


[replay-5.zip](https://github.com/vegaprotocol/vega/files/13347959/replay-5.zip)
[vegahome.zip](https://github.com/vegaprotocol/vega/files/13347968/vegahome.zip)


### Steps to reproduce

```bash
1. In market sim repo run scripts/run-fuzz-test.sh
2. Or Run attached replay file using attached market sim config files
```


### Software version

aa2980058163d7c4d83a0347b762a1a1b1093dc7

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,7 @@
 
 - [9941](https://github.com/vegaprotocol/vega/issues/9941) - Add data node mapping for `WasEligible` field in referral set.
 - [9940](https://github.com/vegaprotocol/vega/issues/9940) - Truncate fee stats in quantum down to the 6 decimal.
+- [9940](https://github.com/vegaprotocol/vega/issues/9940) - Do not assume stop order is valid when generating ids up front.
 - [9998](https://github.com/vegaprotocol/vega/issues/9998) - Slippage factors can now me updated in a market.
 - [10036](https://github.com/vegaprotocol/vega/issues/10036) - Average entry price no longer flickers after a trade.
 - [9956](https://github.com/vegaprotocol/vega/issues/9956) - Prevent validator node from starting if they do not have a Ethereum `RPCAddress` set.

--- a/core/processor/batch_market_instructions_processor.go
+++ b/core/processor/batch_market_instructions_processor.go
@@ -124,20 +124,12 @@ func (p *BMIProcessor) ProcessBatch(
 	// these need to be determinitistic
 	idgen := idgeneration.New(determinitisticID)
 
-	// we will need an ID for every stop orders
-	// not all StopsOrderSubmission have 2 StopOrder
-	stopOrderSize := 0
-	for _, v := range batch.StopOrdersSubmission {
-		if v.FallsBelow != nil {
-			stopOrderSize++
-		}
-		if v.RisesAbove != nil {
-			stopOrderSize++
-		}
-	}
+	// each order will need a new ID, and each stop order can contain up to two orders (rises above, falls below)
+	// but a stop order could also be invalid and have neither, so we pre-generate the maximum ids we might need
+	nIDs := len(batch.Submissions) + (2 * len(batch.StopOrdersSubmission))
 
-	submissionsIDs := make([]string, 0, len(batch.Submissions)+stopOrderSize)
-	for i := 0; i < len(batch.Submissions)+stopOrderSize; i++ {
+	submissionsIDs := make([]string, 0, nIDs)
+	for i := 0; i < nIDs; i++ {
 		submissionsIDs = append(submissionsIDs, idgen.NextID())
 	}
 

--- a/core/processor/batch_market_instructions_processor_test.go
+++ b/core/processor/batch_market_instructions_processor_test.go
@@ -295,3 +295,53 @@ func TestBatchMarketInstructionsEnsureAllErrorReturnNonPartialError(t *testing.T
 	assert.True(t, ok)
 	assert.False(t, perr.IsPartial())
 }
+
+func TestBatchMarketInstructionInvalidStopOrder(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	exec := mocks.NewMockExecutionEngine(ctrl)
+	proc := processor.NewBMIProcessor(logging.NewTestLogger(), exec, processor.Validate{})
+	stats := stats.New(logging.NewTestLogger(), stats.NewDefaultConfig())
+
+	batch := commandspb.BatchMarketInstructions{
+		StopOrdersSubmission: []*commandspb.StopOrdersSubmission{
+			{}, // this one is invalid
+			{
+				RisesAbove: &commandspb.StopOrderSetup{
+					Trigger: &commandspb.StopOrderSetup_Price{Price: "1000"},
+					OrderSubmission: &commandspb.OrderSubmission{
+						MarketId:    "92eea9006eaa51154cb9110b9fe982f37d3bd50f62ee9d0a7709d9c74de329aa",
+						Size:        1,
+						Side:        vega.Side_SIDE_SELL,
+						Type:        vega.Order_TYPE_MARKET,
+						TimeInForce: types.OrderTimeInForceFOK,
+						ReduceOnly:  true,
+					},
+				},
+			},
+		},
+	}
+	stopCnt := 0
+	exec.EXPECT().SubmitStopOrders(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1).DoAndReturn(
+		func(ctx context.Context, stop *types.StopOrdersSubmission, party string, idgen common.IDGenerator, id1, id2 *string) ([]*types.OrderConfirmation, error) {
+			stopCnt++
+			return nil, nil
+		},
+	)
+
+	err := proc.ProcessBatch(
+		context.Background(),
+		&batch,
+		"43f86066fe13743448442022c099c48abbd7e9c5eac1c2558fdac1fbf549e867",
+		"62017b6ae543d2e699f41d37598b22dab025c57ed98ef3c237bb91b948c5f8fc",
+		stats.Blockchain,
+	)
+
+	assert.EqualError(t, err, "0 (* (must have at least one of rises above or falls below))")
+
+	// ensure the errors is reported as partial
+	perr, ok := err.(abci.MaybePartialError)
+	assert.True(t, ok)
+	assert.True(t, perr.IsPartial())
+	assert.Equal(t, 1, stopCnt)
+}

